docs/code/IDataViewDesignPrinciples.md

```diff
@@ -0,0 +1,479 @@
+* Date time, date time zone, and timespan
+* Key types
+* Vector types
+
+The set of standard types will likely be expanded over time.
+
+The IDataView type system is specified in a separate document, *IDataView Type
+System Specification*.
+
+IDataView provides a general mechanism for associating semantic metadata with
+columns, such as designating sets of score columns, names associated with the
+individual slots of a vector-valued column, values associated with a key type
+column, whether a column's data is normalized, etc.
+
+While IDataView schema supports an arbitrary number of columns, it, like most
+schematized data systems, is designed for a modest number of columns,
+typically, limited to a few hundred. When a large number of *features* are
+required, the features should be gathered into one or more vector-valued
+columns, as discussed in the next section. This is important for both user
+experience and performance.
+
+### Vector Types
+
+Machine learning and advanced analytics applications often involve high-
+dimensional data. For example, a common technique for learning from text,
+known as bag-of-words, represents each word in the text as a numeric feature
+containing the number of occurrences of that word. Another technique is
+indicator or one-hot encoding of categorical values, where, for example, a
+text-valued column containing a person's last name is expanded to a set of
+features, one for each possible name (Tesla, Lincoln, Gandhi, Zhang, etc.),
+with a value of one for the feature corresponding to the name, and the
+remaining features having value zero. Variations of these techniques use
+hashing in place of dictionary lookup. With hashing, it is common to use 20
+bits or more for the hash value, producing $2^20$ (about a million) features
+or more.
+
+These techniques typically generate an enormous number of features.
+Representing each feature as an individual column is far from ideal, both from
+the perspective of how the user interacts with the information and how the
+information is managed in the schematized system. The solution is to represent
+each set of features, whether indicator values, or bag-of-words counts, as a
+single vector-valued column.
+
+A vector type specifies an item type and optional dimensionality information.
+The item type must be a primitive, non-vector, type. The optional
+dimensionality information specifies, at the basic level, the number of items
+in the corresponding vector values.
+
+When the size is unspecified, the vector type is variable-length, and
+corresponding vector values may have any length. A tokenization transform,
+that maps a text value to the sequence of individual terms in that text,
+naturally produces variable-length vectors of text. Then, a hashing ngram
+transform may map the variable-length vectors of text to a bag-of-ngrams
+representation, which naturally produces numeric vectors of length $2^k$, where
```

Comment:
I don't think math renders in github. Could hot-link to a service such as: <img src="https://latex.codecogs.com/gif.latex?$2^k$" title="$2^k$" />

src: `<img src="https://latex.codecogs.com/gif.latex?$2^k$" title="$2^k$" />`

Reply:
BTW don't search for "latex image service" at work or with kids in the room... trust me on this.

---
In reply to: [188975188](https://github.com/dotnet/machinelearning/pull/173#discussion_r188975188) [](ancestors = 188975188,188901632)